src/goto-instrument/ada_checks.cpp

```diff
@@ -0,0 +1,182 @@
+  }
+
+  std::string type2string(Ada_Checkt type) const
+  {
+    switch(type)
+    {
+    case Ada_Checkt::Overflow:
+      return "Overflow";
+    case Ada_Checkt::Range:
+      return "Range";
+    case Ada_Checkt::DivByZero:
+      return "Division by zero";
+    }
+    UNREACHABLE;
+  }
+};
+
+void ada_checkst::replace_with(const std::string &replace_option)
+{
+  auto &goto_functions = goto_model.goto_functions;
+  for(auto &goto_function_pair : goto_functions.function_map)
+  {
+    goto_programt &function_body = goto_function_pair.second.body;
+    for(goto_programt::targett instruction = function_body.instructions.begin();
+        instruction != function_body.instructions.end();
+        ++instruction)
+    {
+      if(instruction->is_function_call())
+      {
+        const auto &function_call = instruction->get_function_call();
+        const auto &function = function_call.function();
+
+        const auto &function_name =
+          id2string(to_symbol_expr(function).get_identifier());
+        if(has_prefix(function_name, CPROVER_ADA_PREFIX))
+        {
+          const auto &function_arguments = function_call.arguments();
+          CHECK_RETURN(!function_arguments.empty());
+
+          replace_by_option(
+            *function_arguments.begin(),
+            function_body,
+            instruction,
+            function_name,
+            replace_option);
+        }
+      }
+    }
+  }
+}
+
+void ada_checkst::replace_by_option(
+  const exprt &condition,
+  goto_programt &goto_program,
+  goto_programt::targett instruction,
+  const std::string &function_name,
+  const std::string &replace_option)
+{
+  exprt boolean_condition = condition;
+  while(boolean_condition.id() == ID_typecast)
+    boolean_condition = to_typecast_expr(boolean_condition).op();
+  CHECK_RETURN(boolean_condition.type().id() == ID_bool);
+
+  source_locationt commented_source_location{};
+  commented_source_location.set_comment(
+    "Ada " + type2string(get_check_type(function_name)) + " check");
+
+  switch(get_replace_type(replace_option))
+  {
+  case Replace_Typet::AssertThenAssume:
+  {
+    auto after_assert = goto_program.insert_after(
+      instruction,
+      goto_programt::make_assertion(
+        boolean_condition, commented_source_location));
+    goto_program.insert_after(
+      after_assert,
+      goto_programt::make_assumption(
+        boolean_condition, commented_source_location));
+    break;
+  }
+  case Replace_Typet::AssertOnly:
+    goto_program.insert_after(
+      instruction,
+      goto_programt::make_assertion(
+        boolean_condition, commented_source_location));
+    break;
+  case Replace_Typet::AssumeOnly:
+    goto_program.insert_after(
+      instruction,
+      goto_programt::make_assumption(
+        boolean_condition, commented_source_location));
+    break;
+  }
+
+  auto skip_instruction = goto_programt::make_skip();
+  instruction->swap(skip_instruction);
```

Comment:
Why is this happening here? 

(I'm not saying it's wrong, I just don't understand what purpose it serves)

Reply:
Well the idea is that we _replace_ the original call to `__CPROVER_Ada_check` with assert/assume. On the lines above we have added the assertion, and here we change the original function call to a `SKIP`. The only interface for changing an existing instruction I could find is `swap`, so that's what I used.